akka-docs/rst/java/code/docs/http/javadsl/server/directives/SecurityDirectivesExamplesTest.java

```diff
@@ -0,0 +1,361 @@
+      .assertHeaderExists("WWW-Authenticate", "MyAuth realm=\"MyRealm\"");
+
+    final HttpCredentials validCredentials =
+      BasicHttpCredentials.createBasicHttpCredentials("John", "p4ssw0rd");
+    testRoute(route).run(HttpRequest.GET("/secured").addCredentials(validCredentials))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("Authenticated!");
+    //#authenticateOrRejectWithChallenge
+  }
+
+  @Test
+  public void testAuthorize() {
+    //#authorize
+    class User {
+      private final String name;
+      public User(String name) {
+        this.name = name;
+      }
+      public String getName() {
+        return name;
+      }
+    }
+
+    // authenticate the user:
+    final Function<Optional<ProvidedCredentials>, Optional<User>> myUserPassAuthenticator =
+      opt -> {
+        if (opt.isPresent()) {
+          return Optional.of(new User(opt.get().identifier()));
+        } else {
+          return Optional.empty();
+        }
+      };
+
+    // check if user is authorized to perform admin actions:
+    final Set<String> admins = new HashSet<>();
+    admins.add("Peter");
+    final Function<User, Boolean> hasAdminPermissions = user -> admins.contains(user.getName());
+
+    final Route route = authenticateBasic("secure site", myUserPassAuthenticator, user ->
+      path("peters-lair", () ->
+        authorize(() -> hasAdminPermissions.apply(user), () ->
+          complete("'" + user.getName() +"' visited Peter's lair")
+        )
+      )
+    ).seal(system(), materializer());
+
+    // tests:
+    final HttpCredentials johnsCred =
+      BasicHttpCredentials.createBasicHttpCredentials("John", "p4ssw0rd");
+    testRoute(route).run(HttpRequest.GET("/peters-lair").addCredentials(johnsCred))
+      .assertStatusCode(StatusCodes.FORBIDDEN)
+      .assertEntity("The supplied authentication is not authorized to access this resource");
+
+    final HttpCredentials petersCred =
+      BasicHttpCredentials.createBasicHttpCredentials("Peter", "pan");
+    testRoute(route).run(HttpRequest.GET("/peters-lair").addCredentials(petersCred))
+      .assertEntity("'Peter' visited Peter's lair");
+    //#authorize
+  }
+
+  @Test
+  public void testAuthorizeAsync() {
+    //#authorizeAsync
+    class User {
+      private final String name;
+      public User(String name) {
+        this.name = name;
+      }
+      public String getName() {
+        return name;
+      }
+    }
+
+    // authenticate the user:
+    final Function<Optional<ProvidedCredentials>, Optional<User>> myUserPassAuthenticator =
+      opt -> {
+        if (opt.isPresent()) {
+          return Optional.of(new User(opt.get().identifier()));
+        } else {
+          return Optional.empty();
+        }
+      };
+
+    // check if user is authorized to perform admin actions,
+    // this could potentially be a long operation so it would return a Future
+    final Set<String> admins = new HashSet<>();
+    admins.add("Peter");
```

Comment:
Since this is mutable it is not safe to access it inside of the completable future, needs to be immutable, or a threadsafe set.